api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+        if (userCanVoteOnContract(voter, contractModel)) {
+            final VoteModel voteModel = VoteModel.builder()
+                    .voter(voter)
+                    .contract(contractModel)
+                    .viewerFlaggedComplete(flaggedCompleted)
+                    .votedAt(new Timestamp(System.currentTimeMillis()))
+                    .build();
+            voteModelRepository.save(voteModel);
+        }
+    }
+
+    @Override
+    public boolean isVotingComplete(@NonNull final ContractModel contractModel) {
+        validateContract(contractModel);
+
+        final UUID contractId = contractModel.getId();
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getProposer().getId());
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getStreamer().getId());
+        final Timestamp now = new Timestamp(System.currentTimeMillis());
+        return (proposerVote != null && streamerVote != null) ||
+               (proposerMarkedContractCompleted(contractModel)) ||
+               (streamerMarkedContractFailed(contractModel)) ||
+               (now.after(contractModel.getExpiresAt()));
+    }
+
+    @Override
+    public boolean wasContractCompleted(@NonNull final ContractModel contractModel) {
+        if (!isVotingComplete(contractModel)) {
+            throw new IllegalArgumentException("Attempted to check if contract was completed before voting was finished. ContractId: " + contractModel.getId());
+        }
+
+        final UUID contractId = contractModel.getId();
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getProposer().getId());
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getStreamer().getId());
+        if (proposerMarkedContractCompleted(contractModel)) {
+            //Proposer has marked contract as completed. Release funds without reading Streamer's vote.
+            return true;
+        } else if (streamerMarkedContractFailed(contractModel)) {
+            //Streamer has marked contract as failed. Return funds without reading Proposers's vote.
+            return false;
+        } else if(proposerVote == null && streamerVote == null) {
+            //Neither proposer or streamer voted. Release funds if contract was accepted. Return funds otherwise.
+            return contractModel.isAccepted();
+        } else if(proposerVote != null) {
+            //Streamer did not vote. Take Proposer's vote as the final outcome
+            return proposerVote.isViewerFlaggedComplete();
+        } else if(streamerVote != null){
+            //Proposer did not vote. Take Streamer's vote as the final outcome
+            return streamerVote.isViewerFlaggedComplete();
+        }
+        //Default case is to consider the contract completed and release money to the streamer.
```

Comment:
Unless I'm mistaken, this is the case where both the proposer and the streamer have vote, but there's a disagreement right? We should definitely log this case, perhaps even do something like send an email to us.

Reply:
Yeah this is for the disagreement case. Im with you on logging. Ill add that for now and we can discuss sending emails on our next call